### DISCUSSION TITLE
클래스 안에서의 private method, private property, public, setter와 getter, static 등의 선언 순서

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-code-review-study#23] [Lotto]
클래스 안에서 private와 public 프로퍼티와 메서드와 static 메서드 등이 동시에 존재할 때
어떤 순서로 선언을 해주는 것이 맞을까요? 또 setter는 getter와 붙어 있는 것이 맞을까요?

**Comment by lokba** [comment]
오호라,,신선하네요.


저는 이런식으로 구성할 것 같아요.

```
class A {
  set Name() {}

  get Name() {}

  #a() {}

  #b() {}

  static e() {}

  c() {}
  d() {}
}

```

이유는 다음과 같습니다.
리뷰어분께서 `public과 private 메소드가 너무 뒤섞여있다.`라는 코멘트를 봤기에, 우선 크게 private -> public순으로 구성했습니다.
setter,getter은 직접적으로 상태에 접근하는 로직이 담겨있기에, 최대한 상태값 관련 코드는 보통 클래스 최상단에 위치해 있기에 #메소드보다 위에 배치했습니다.

static  유/무 메소드는 사실 상황마다 위치가 좀 다를것 같습니다. 우선은 굳이 정한다면, static 메소드를 더 상단에 위치할 것 같습니다.

**Comment by wonsss** [comment]
책 클린코드 10장 클래스에서 클래스 체계에 대해 다음과 같이 설명하고 있어요.
> 클래스를 정의하는 표준 자바 관례에 따르면, 가장 먼저 `변수 목록`이 나온다. `static public 상수`가 있다면 맨 처음에 나온다. 다음으로 `static private 변수`가 나오며, 이어서 `private 인스턴스 변수`가 나온다. `public 변수`가 필요한 경우는 거의 없다.  
변수 목록 다음에는 `public 함수`가 나온다. `private 함수`는 자신을 호출하는 공개 함수 직후에 넣는다. 
즉, 추상화 단계가 순차적으로 내려간다. 
그래서 프로그램은 `신문 기사`처럼 읽힌다.

이를 다시 정리하자면
1. static public 상수
2. static private 변수
3. private instance 변수(public instance 변수가 필요한 경우는 거의 없음)
4. public method
   - private method는 자신을 호출하는 public method 직후에 위치
5. 그 다음 public method ... 계속 

이는 자바 기준이라서, 자바스크립트에서는 정확히 어떻게 해석되는지는 잘 모르겠어요.
소속된 팀에서 어떤 순서가 가독성에 좋은지 컨벤션으로 결정할 문제이므로 절대적인 것은 없다고 생각돼요.
증요한 것은 신문기사처럼 위에서부터 아래로 논리적으로 읽히냐인 것 같아요.

일단, 저는 아래와 같은 순서가 괜찮을 것 같은데, 어떠신가요?

기준으로는 public -> 해당 private

애매한 것은 constructor의 위치와 getter, setter의 위치입니다. 
그러나 먼저 멤버변수를 선언한 후에 초기화를 해줘야 할 것 같았고, getter나 setter도 크게 public 메서드라고 볼 수 있으니까
변수 선언한 다음 순서로 잡아줬어요.

```js
class A {
	#myPrivateMember1;  // private instance 변수
	#myPrivateMember2;  

	constructor() {       // 초기화
	}

	get myPrivateMember1() {    // getter와 setter들은 모두 같이 모아둔다.
	}

	set myPrivateMember1() {
    }

	myPublicMethod1() {         // public 메서드
			this.#myPrivateMehtod1();     
	}
	
    #myPrivateMethod1() {     // public에서 호출하는 private 메서드를 바로 다음에 선언
    }

	myPublicMethod2() {         // 다른 public 메서드
			this.#myPrivateMehtod2();     
	}
	
    #myPrivateMethod2() {
    }

	static myStaticMethod() {   // static 메서드는 객체와 연관성이 제일 떨어지니까 마지막에 선언하거나 모듈화해서 extends해서 사용
	}
```

P.S.
아 그리고 찾다보니까 구글에서는 가독성에 문제가 없다면 static method들을 모듈화하여 extends해서 사용하는 방식을 따르고 있네요.
https://softwareengineering.stackexchange.com/questions/240968/should-our-team-order-javascript-class-methods-properties-consistently-if-so-h

또한, 구글은 getter랑 setter를 안 쓰고 대신 일반적인 메서드를 사용하는 것 같아요. https://google.github.io/styleguide/jsguide.html#features-classes-getters-and-setters
